core/src/main/java/org/infinispan/interceptors/locking/AbstractTxLockingInterceptor.java

```diff
@@ -56,20 +57,18 @@ public Object visitRollbackCommand(TxInvocationContext ctx, RollbackCommand comm
    }
 
    @Override
-   public Object visitGetKeyValueCommand(InvocationContext ctx, GetKeyValueCommand command) throws Throwable {
-      try {
-         return super.visitGetKeyValueCommand(ctx, command);
-      } finally {
-         //when not invoked in an explicit tx's scope the get is non-transactional(mainly for efficiency).
-         //locks need to be released in this situation as they might have been acquired from L1.
-         if (!ctx.isInTxScope()) lockManager.unlockAll(ctx);
+   public Object visitPutKeyValueCommand(InvocationContext ctx, PutKeyValueCommand command) throws Throwable {
+      if (command.hasFlag(Flag.PUT_FOR_EXTERNAL_READ)) {
+         // Cache.putForExternalRead() is non-transactional
+         return visitNonTxDataWriteCommand(ctx, command);
       }
+      return visitDataWriteCommand(ctx, command);
    }
 
    @Override
-   public Object visitGetCacheEntryCommand(InvocationContext ctx, GetCacheEntryCommand command) throws Throwable {
+   protected Object visitDataReadCommand(InvocationContext ctx, DataCommand command) throws Throwable {
```

Comment:
It seems this is only used in OptimisticLockingInterceptor as Pessimistic doesn't call super so I would say to just move this method into Optimistic directly.